core/src/main/java/hudson/model/User.java

```diff
@@ -370,21 +375,36 @@ public synchronized void addProperty(@NonNull UserProperty p) throws IOException
     }
 
     /**
-     * This method checks with {@link SecurityRealm} if the user is a valid user that can login to the security realm.
-     * If {@link SecurityRealm} is a kind that does not support querying information about other users, this will
-     * use {@link LastGrantedAuthoritiesProperty} to pick up the granted authorities as of the last time the user has
-     * logged in.
+     * This method checks with {@link UserDetailsCache} and {@link SecurityRealm} if the user is a valid user that can 
+     * login to the security realm. If {@link SecurityRealm} is a kind that does not support querying information about
+     * other users, this will use {@link LastGrantedAuthoritiesProperty} to pick up the granted authorities as of the 
+     * last time the user has logged in.
      *
      * @return userDetails for the user, in case he's not found but seems legitimate, we provide a userDetails with minimum access
      * @throws UsernameNotFoundException If this user is not a valid user in the backend {@link SecurityRealm}.
      */
     public @NonNull UserDetails getUserDetailsForImpersonation() throws UsernameNotFoundException {
-        ImpersonatingUserDetailsService userDetailsService = new ImpersonatingUserDetailsService(
-                Jenkins.get().getSecurityRealm().getSecurityComponents().userDetails
-        );
-
         try {
-            UserDetails userDetails = userDetailsService.loadUserByUsername(id);
+            UserDetails userDetails;
+            if (!DISABLE_CACHE_FOR_IMPERSONATION) {
+                try {
+                    userDetails = UserDetailsCache.get().loadUserByUsername(id);
+                } catch (ExecutionException ex) {
+                    LOGGER.log(Level.FINE, "Failure to retrieve " + id + "from cache", ex);
+                    userDetails = new ImpersonatingUserDetailsService(
+                        Jenkins.get().getSecurityRealm().getSecurityComponents().userDetails
+                    ).loadUserByUsername(id);
+                } catch (DataAccessException | UserMayOrMayNotExistException e) {
+                    LOGGER.log(Level.FINE, "Failure to retrieve " + id , e);
+                    // Those Exception originates from the SecurityRealm and tell us that the User may or may not exists
+                    // So directly try to impersonate
+                    userDetails = Optional.ofNullable(getUserDetailsBase()).orElseThrow(() -> e);
+                }
```

Comment:
If we hit `DataAccessException` / `UserMayOrMayNotExistException` when retrieving from the `UserDetailsCache`, we know those come from the `SecurityRealm`. If we call the `ImpersonatingUserDetailsService` in that case, we will have the same result, but that will waste another call to the `SecurityRealm#loadUserByUsername`. Saving that call is the reason for the nested try / catch.
I see you point with the DRY code. I can improve this with your suggestion.

Reply:
And with remote calls, that could be annoying in terms of performance. My goal is to prevent two different codes, doing the stuff to be called (exception management in this case), so if you can propose a way that is not calling the security realm twice, it could be better :)